OpenSim/Common/Test/testDataInterface.cpp

```diff
@@ -0,0 +1,375 @@
+        _inTable->appendRow(row);
+        return !_inputstream->eof();
+    }
+
+private:
+    SimTK::ReferencePtr<std::istream> _inputstream;
+    mutable SimTK::ReferencePtr<DataTable_<DataType> > _inTable;
+};
+
+using namespace std;
+using namespace SimTK;
+
+int main() {
+
+    try {
+        DataAdapter::registerDataAdapter("streamWriter",
+            StreamWriter<SimTK::Real>());
+        DataAdapter::registerDataAdapter("csv", CSVFileReader());
+
+        // create a data table of Vec3's, which would be like marker data
+        DataTable_<Vec3> markers(1,4);
+        RowVector_<Vec3> r0(4, Vec3(1.0, 1.0, 1.0));
+        RowVector_<Vec3> r1(4, Vec3(2.0, 2.0, 2.0));
+        RowVector_<Vec3> r2(4, Vec3(3.0, 3.0, 3.0));
+        
+        // Update the table by row
+        markers.updRow(0) = r0;
+        // TODO updating the next row should throw an exception since markers
+        // should only have 1 row. Currently aborts in release. SimTK::Matrix
+        // checks in debug.
+        //ASSERT_THROW(std::exception, markers.updRow(1) = r1);
+
+        // Append more rows to grow the table.
+        markers.appendRow(r1);
+        markers.appendRow(r2);
+        cout << "HERE 2" << endl;
+        ASSERT_THROW(std::exception,
+             markers.appendRow(RowVector_<Vec3>(3, Vec3(4.0, 4.0, 4.0))) );
+
+        // create and update column labels for the table of marker data
+        Array<string> labels;
+        labels.append("mark1"); labels.append("mark2"); 
+        labels.append("mark3"); labels.append("mark4");
+        markers.updColumnLabels() = labels;
+
+        // demonstrate row and column access of the data table
+        size_t nr = markers.getNumRows();
+        size_t nc = markers.getNumCols();
+
+        cout << "Markers has " << nr << " rows, " << nc << " columns." << endl;
+        cout << "Marker columns are " << markers.getColumnLabels() << endl;
+
+        const RowVector_<Vec3> row = markers.getRow(2);
+        row.dump("Row 2 markers");
+
+        const Vector_<Vec3> column = markers.getColumn(3);
+        column.dump("Column 3 marker");
+
+        cout << "Element(1,3) = " << markers.getElement(1, 3) << endl;
+
+        cout << "Markers Table:" << endl;
+        // exercise DataTable's output operator
+        cout << markers << endl;
+
+        // DataTable is our typical table of Real numbers
+        DataTable table;
+        // Build up the table by row
+        table.appendRow(RowVector(5, 0.1));
+        table.appendRow(RowVector(5, 0.2));
+        table.appendRow(RowVector(5, 0.3));
+
+        nr = table.getNumRows();
+        nc = table.getNumCols();
+
+        cout << "\nTable of values has " << nr << " rows, " << nc << " columns." << endl;
+
+        labels.setSize(0);
+        labels.append("time");
+        for (int i = 1; i < nc; ++i){
+            stringstream label;
+            label << "col" << i;
+            labels.append(label.str());
+        }
+
+        table.updColumnLabels() = labels;
```

Comment:
This might not be an obvious way for many less-fluent C++ users to update all the labels.